snapcraft/internal/lifecycle.py

```diff
@@ -243,6 +244,18 @@ def _run_step(self, step, part, part_names):
         common.env.extend(self.config.project_env())
 
         part = _replace_in_part(part)
+
+        # Add the annotated list of build packages
+        part_build_packages = part._part_properties['build-packages']
+        part.build_packages = []
+        for pkg in self.build_packages:
```

Comment:
self.build_packages is the list of *all* the build packages.  The loop is to get only the versioned package that are used by this part.